### PR TITLE
Show when user can't connect because of a bad token

### PR DESCRIPTION
**Pull request by lf94**
This PR also resolves a few other connection related issues:

* Revert back to the new event-based engine network events.
* Use Context API to properly create a application-wide network status.
* Fix a couple errors related to assumptions in network state.
* And of course the mentioned showing when a user is trying to connect with a bad token.

This is a preliminary PR because there are other issues that need to be addressed before we can improve the experience.

The other issues are related to api-deux:

* https://github.com/KittyCAD/api-deux/issues/1629
* https://github.com/KittyCAD/api-deux/issues/1628

I can work on these when I'm home Wednesday. Otherwise I would prefer someone who deals with api-deux often to deal with these after this PR is merged.

Until then, I will continue with connection related issues, like "the concept of idle": https://github.com/KittyCAD/modeling-app/issues/563, and "more than 2 connections at once": https://github.com/KittyCAD/modeling-app/issues/1198

closes https://github.com/KittyCAD/modeling-app/issues/1561

### CODE REVIEW

File: src/lang/std/engineConnection.ts [hunk 1]
```diff
@@ -210,99 +268,76 @@ class EngineConnection {
...
       }
     }
     this._state = next
-    this.onConnectionStateChange(this._state)
+
+    this.dispatchEvent(
+      new CustomEvent(EngineConnectionEvents.ConnectionStateChanged, {
+        detail: this._state,
+      })
+    )
```
Comment (Irev-Dev) [inline src/lang/std/engineConnection.ts:290]: I don't love using such a browser-specific feature, as eventually, we'll want to migrate as much of this as possible rust/wasm. That was the motivation for not integrating this into Xstate/the rest of the app's state management and using callbacks because class and callbacks are pretty generic and will much more easily convert to rust.

I get this is a vague hypothetical without a date on when we'll carry out this work 🤷 

What was the motivation for using Events again? (I say again because @paultag gravitated towards them too).

File: src/editor/plugins/lsp/copilot/index.ts
[the remark this answers is not in the document]
Comment (jessfraz) [reply src/editor/plugins/lsp/copilot/index.ts:1]: Looks like nothing is authed anymore

[the remark this answers is not in the document]
Comment (jessfraz) [reply src/editor/plugins/lsp/copilot/index.ts:1]: Specifically the lsp servers

File: src/lang/std/engineConnection.ts (hunk 1 above)
Comment (lf94) [reply src/lang/std/engineConnection.ts:290]: The motivation for using events is we were already using events for this before, but instead this uses the Events API instead of a new events API implementation. We already had instances of the code creating subscribers.

Additionally, it makes sense to use events as a programming method here because many of the objects we communicate also use this event-based paradigm. Having a paradigm change creates a weird interface we have to maintain and further think about how everything maps or connects together.

The Events API is not browser specific: https://developer.mozilla.org/en-US/docs/Web/API/Event#browser_compatibility you can see Deno and NodeJS support this API.

Rust has many event based packages, I'm not sure why it would matter if we used a callback based vs event based API:

* https://crates.io/crates/presage
* https://crates.io/crates/eventify
* https://crates.io/crates/event
* https://crates.io/crates/statig

Additionally even if we want move engineConnection into Rust, it may not be the best idea, because there are many browser related mechanisms we use already.

We may already be past the point of "we want it all in Rust". It's very possible the future of this looks like a complete rewrite.

I hope that clarifies why I've gone with the Events API.

Comment (lf94) [reply src/lang/std/engineConnection.ts:290]: As I'm sitting here at the airport I think I've thought of further "why".

Callbacks are usually for a specific call-site and only one can be bound.

Events can be subscribed to from anywhere and many can be bound.

We can probably "downgrade" to callbacks (I don't mean this as a bad thing) with some careful thinking to make this _perfect_.

There is some level of simultaneous operations that are going on in engineConnection, but I also understand we actually care about mostly sequencing it instead. This would be better to address during a rewrite to Rust.

I would further say that writing JS/TS in an idiomatic way is much easier to maintain than in a Rust idiomatic way _in JS/TS_ because there will be concepts which are weird to express in one or the other. (Just look at our C++-Rust bridge which are arguably closer to eachother  in capabilities...)

File: src/editor/plugins/lsp/copilot/index.ts
[the remark this answers is not in the document]
Comment (lf94) [reply src/editor/plugins/lsp/copilot/index.ts:1]: It might actually be the catching of the lsp promise I added elsewhere in this PR.

[the remark this answers is not in the document]
Comment (lf94) [reply src/editor/plugins/lsp/copilot/index.ts:1]:
```ts
        lspServer
          .start('kcl', token)
          .then(() => {
            setIsKclLspServerReady(true)
          })
          .catch(console.warn)
```

**Comment by jessfraz** [comment]
I think this fixes https://github.com/KittyCAD/modeling-app/issues/1770

**Comment by jessfraz** [comment]
We should be able to test this in tauri e2e on windows when set up just adding a note so later when I search I find this

File: e2e/playwright/flow-tests.spec.ts [hunk 2]
```diff
@@ -2967,6 +2966,30 @@ const part002 = startSketchOn('XZ')
...
+    const [x1, y1] = [0, Math.sin((tau / steps) * j * freq) * amplitude]
+    const [x2, y2] = [
+      Math.cos(-ang * deg) * i - Math.sin(-ang * deg) * y1,
+      Math.sin(-ang * deg) * i + Math.cos(-ang * deg) * y1,
+    ]
+    const [xr, yr] = [x2, y2]
+    await page.mouse.move(x + xr, y + yr, { steps: 2 })
+  }
+}
+
```
Comment (franknoirot) [inline e2e/playwright/flow-tests.spec.ts:2992]: ❤️

File: e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png
Comment (franknoirot) [inline e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png:1]: All these snapshots have the engine-side gizmo in them: is it possible that the gizmo code is still around after a rebase/merge?

File: src/hooks/useSetupEngineManager.ts [hunk 3]
```diff
@@ -43,7 +43,7 @@ export function useSetupEngineManager(
     engineCommandManager.pool = settings.pool
   }
 
-  useLayoutEffect(() => {
+  const startEngineInstance = () => {
```
Comment (franknoirot) [inline src/hooks/useSetupEngineManager.ts:46]: I'm not certain of the performance implications: should this use [`useCallback` to cache the callback between rerenders](https://react.dev/reference/react/useCallback)?

File: src/lang/std/engineConnection.ts [hunk 4]
```diff
@@ -1179,23 +1331,36 @@ export class EngineCommandManager {
...
+                  type: EngineConnectionStateType.Disconnecting,
+                  value: {
+                    type: DisconnectingType.Error,
+                    value: {
+                      error: ConnectionError.LostVideoStream,
+                    },
+                  },
+                }
+              }
+            })
```
Comment (franknoirot) [inline src/lang/std/engineConnection.ts:1353]: Muting the stream causes a lost video stream error?

**Review by franknoirot — COMMENTED** [review]
Nice! Definitely some stuff I don't understand here at a glance, but looks good as far as I can tell aside from the fact that the engine gizmo is appearing in the snapshots

File: src/lang/std/engineConnection.ts (hunk 4 above)
Comment (lf94) [reply src/lang/std/engineConnection.ts:1353]: This logic was already there before; it's an early signal that the stream is lost (the video goes mute). I just added the event 'sall. It's actually pretty effective way to see an early stream close.

File: src/hooks/useSetupEngineManager.ts (hunk 3 above)
Comment (lf94) [reply src/hooks/useSetupEngineManager.ts:46]: This was part of a rebase; someone else added this :shrug:

File: e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png
Comment (lf94) [reply e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png:1]: It's possible, or it was not part of my rebase? Not sure. I don't remember touching anything gizmo related...

File: src/hooks/useSetupEngineManager.ts (hunk 3 above)
Comment (franknoirot) [reply src/hooks/useSetupEngineManager.ts:46]: Yeah absolutely, no worries

File: e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png
Comment (franknoirot) [reply e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png:1]: Maybe it was an earlier PR like mine, I'll have to search the code base for references to "gizmo" real quick

Comment (franknoirot) [reply e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png:1]: Nah it's just in this PR [here](https://github.com/KittyCAD/modeling-app/blob/connection-20240413/src/lang/std/engineConnection.ts#L1218-L1235) and maybe a couple other places. Searching for references to "make_axes_gizmo" is effective

Comment (franknoirot) [reply e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png:1]: All references to the engine side gizmo were removed as a part of #2354

Comment (lf94) [reply e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png:1]: I botched some rebase then right?

Comment (franknoirot) [reply e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png:1]: Botched is too strong a word, but yeah some of your changes in engineConnection must have clobbered some that came in from main, so far all I see is the gizmo removal

Comment (lf94) [reply e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png:1]: I frickin' spaghetti mixed it Frank! :p

File: src/Router.tsx [hunk 5]
```diff
@@ -155,5 +157,11 @@ const router = createBrowserRouter([
  * @returns RouterProvider
  */
 export const Router = () => {
-  return <RouterProvider router={router} />
+  const networkStatus = useNetworkStatus()
+
+  return (
+    <NetworkContext.Provider value={networkStatus as any}>
```
Comment (Irev-Dev) [inline src/Router.tsx:163]: Can this by typed?

**Comment by lf94** [comment]
@Irev-Dev oop! Nice catch. I only tested against our dev servers and not a local engine, so I'll check this out tomorrow (Monday) :slightly_smiling_face: Thank you!

File: src/Router.tsx (hunk 5 above)
Comment (lf94) [reply src/Router.tsx:163]: Should be no problem.

**Comment by Irev-Dev** [comment]
> @Irev-Dev oop! Nice catch. I only tested against our dev servers (and I suppose prod servers since that's what our CI tests run against) and not a local engine, so I'll check this out tomorrow (Monday) 🙂 Thank you!

Ah k, yeah don't know why it would be different, and to be fair it does eventually work, sorry I should't have also tried you branch against prod to see if the wait was there too.

Sorry something else I wanted to ask about was moving the test utils into `e2e/playwright/test-utils.ts` for a couple reasons
- makes it consistent with other test utils
- would be available to `e2e/playwright/snapshot-tests.spec.ts` as well if we wanted to use them there
- And while not your fault there seems to be a problem with mixing normal variable and function declarations between `test(...`s (other than up the top of the module) and the integration with the vscode extension. (most of the test would disappear in the extension, can still run them with the cli, but given I do find the extension pretty useful I'd rather not break it)

I didn't want to ask you to do yet another thing so I did try with this
https://github.com/KittyCAD/modeling-app/pull/2574
It's aimed at main because it needs to be to trigger the playw tests, but you can see the [diff against this branch here](https://github.com/KittyCAD/modeling-app/compare/connection-20240413...kurt-connection-20240413), but I'm having trouble with macos failing (timing out rather than failing really)

And related to the above, I noticed that the last commit that's got all the ✅ s, 08c1745be093fe9fdd68cdbfc5bcdc9cc893074d that macos had to be run three times. I'm happy for some of this stuff be fast follows btw, but as you know failing tests are really annoying.

Maybe it's still pragmatic to keep the workers to 1 if that makes these tests slightly better for the time being?

**Comment by lf94** [comment]
Thanks man, I'll sort it all in the morning :)

File: e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png
Comment (lf94) [reply e2e/playwright/snapshot-tests.spec.ts-snapshots/Client-side-scene-scale-should-match-engine-scale-Inch-scale-2-Google-Chrome-linux.png:1]: Phew, turns out this was just a single engine command I didn't remove.

**Review by franknoirot — APPROVED** [review]
I've confirmed that the new utils have been moved to test-utils

File: playwright.config.ts [hunk 6]
```diff
@@ -72,7 +72,7 @@ export default defineConfig({
 
   /* Run your local dev server before starting the tests */
   webServer: {
-    command: 'yarn serve',
+    command: 'yarn start',
```
Comment (Irev-Dev) [inline playwright.config.ts:75]: Why this change, does it make a difference?

Would be slightly more comfortable testing against "serve" since that's closer to the prod setup, but I guess if it makes tests more reliable, I want that even more.

**Review by Irev-Dev — APPROVED** [review]

File: playwright.config.ts (hunk 6 above)
Comment (lf94) [reply playwright.config.ts:75]: No, it makes no difference from my experience. I will change this back in a follow up
